File: src/queries.ts
import type { TerraformConfig, TfBlock } from "./types.js";
import type { EventLog } from "@aetherwing/fcp-core";
import type { TerraformEvent } from "./types.js";
import { serializeToHcl } from "./hcl.js";
import { findByLabel, findByKind } from "./model.js";

export function dispatchQuery(
  query: string,
  config: TerraformConfig,
  eventLog?: EventLog<TerraformEvent>,
): string {
  const parts = query.trim().split(/\s+/);
  const cmd = parts[0]?.toLowerCase() ?? "";
  const args = parts.slice(1);

  switch (cmd) {
    case "map":
      return queryMap(config);
    case "list":
      return queryList(config, args);
    case "describe":
      return queryDescribe(config, args[0]);
    case "plan":
      return queryPlan(config);
    case "graph":
      return queryGraph(config);
    case "validate":
      return queryValidate(config);
    case "stats":
      return queryStats(config);
    case "status":
      return queryStatus(config);
    case "find":
      return queryFind(config, args.join(" "));
    case "history":
      return queryHistory(eventLog, parseInt(args[0]) || 10);
    default:
      return `Unknown query: "${cmd}". Available: map, list, describe, plan, graph, validate, stats, status, find, history`;
  }
}

function queryMap(config: TerraformConfig): string {
  const lines: string[] = [];
  lines.push(`Terraform Config: ${config.title}`);
  lines.push("");

  const resources = findByKind(config, "resource");
  const variables = findByKind(config, "variable");
  const outputs = findByKind(config, "output");
  const providers = findByKind(config, "provider");
  const data = findByKind(config, "data");

  if (providers.length > 0) {
    lines.push(`Providers: ${providers.map((p) => p.label).join(", ")}`);
  }

  if (resources.length > 0) {
    const typeCounts = new Map<string, number>();
    for (const r of resources) {
      typeCounts.set(r.fullType, (typeCounts.get(r.fullType) ?? 0) + 1);
    }
    const types = [...typeCounts.entries()].map(([t, c]) => `${t} x${c}`).join(", ");
    lines.push(`Resources (${resources.length}): ${types}`);
  }

  if (data.length > 0) lines.push(`Data Sources: ${data.length}`);
  if (variables.length > 0) lines.push(`Variables: ${variables.length}`);
  if (outputs.length > 0) lines.push(`Outputs: ${outputs.length}`);
  if (config.connections.size > 0) lines.push(`Connections: ${config.connections.size}`);

  return lines.join("\n");
}

function queryList(config: TerraformConfig, args: string[]): string {
  let blocks = [...config.blocks.values()];

  // Filter by selector if provided
  if (args.length > 0 && args[0].startsWith("@")) {
    const sel = args[0];
    if (sel.startsWith("@type:")) {
      const type = sel.slice(6);
      blocks = blocks.filter((b) => b.fullType === type);
    } else if (sel.startsWith("@kind:")) {
      const kind = sel.slice(6);
      blocks = blocks.filter((b) => b.kind === kind);
    } else if (sel.startsWith("@provider:")) {
      const provider = sel.slice(10);
      blocks = blocks.filter((b) => b.provider === provider);
    }
  }

  if (blocks.length === 0) return "No blocks found.";

  const lines: string[] = [];
  for (const block of blocks) {
    const type = block.kind === "resource" || block.kind === "data"
      ? `${block.fullType}.${block.label}`
      : `${block.kind}.${block.label}`;
    const attrCount = block.attributes.size;
    lines.push(`  ${block.kind.padEnd(10)} ${type.padEnd(35)} (${attrCount} attrs)`);
  }
  return lines.join("\n");
}

function queryDescribe(config: TerraformConfig, label: string | undefined): string {
  if (!label) return "describe requires a LABEL";
  const block = findByLabel(config, label);
  if (!block) return `block "${label}" not found`;

  const lines: string[] = [];
  const ref = block.kind === "resource" || block.kind === "data"
    ? `${block.fullType}.${block.label}`
    : `${block.kind} "${block.label}"`;
  lines.push(`${block.kind}: ${ref}`);
  if (block.provider) lines.push(`  provider: ${block.provider}`);

  if (block.attributes.size > 0) {
    lines.push("  attributes:");
    for (const attr of block.attributes.values()) {
      lines.push(`    ${attr.key} = ${attr.valueType === "string" ? `"${attr.value}"` : attr.value} (${attr.valueType})`);
    }
  }

  if (block.tags.size > 0) {
    lines.push("  tags:");
    for (const [k, v] of block.tags) {
      lines.push(`    ${k} = "${v}"`);
    }
  }

  if (block.nestedBlocks.length > 0) {
    lines.push("  nested blocks:");
    for (const nested of block.nestedBlocks) {
      lines.push(`    ${nested.type} {}`);
      for (const attr of nested.attributes.values()) {
        lines.push(`      ${attr.key} = ${attr.value}`);
      }
    }
  }

  // Show connections
  const conns = [...config.connections.values()].filter(
    (c) => c.sourceId === block.id || c.targetId === block.id,
  );
  if (conns.length > 0) {
    lines.push("  connections:");
    for (const c of conns) {
      if (c.sourceId === block.id) {
        lines.push(`    → ${c.targetLabel}`);
      } else {
        lines.push(`    ← ${c.sourceLabel}`);
      }
    }
  }

  if (block.meta.count) lines.push(`  count: ${block.meta.count}`);
  if (block.meta.forEach) lines.push(`  for_each: ${block.meta.forEach}`);
  if (block.meta.dependsOn.length > 0) lines.push(`  depends_on: ${block.meta.dependsOn.join(", ")}`);

  return lines.join("\n");
}

function queryPlan(config: TerraformConfig): string {
  if (config.blocks.size === 0) return "Empty configuration. Add some resources first.";
  return serializeToHcl(config);
}

function queryGraph(config: TerraformConfig): string {
  if (config.connections.size === 0) return "No connections. Use 'connect SRC -> TGT' to add dependencies.";

  const lines: string[] = ["Dependency Graph:"];
  for (const conn of config.connections.values()) {
    const src = config.blocks.get(conn.sourceId);
    const tgt = config.blocks.get(conn.targetId);
    if (src && tgt) {
      const label = conn.label ? ` (${conn.label})` : "";
      lines.push(`  ${src.label} -> ${tgt.label}${label}`);
    }
  }
  return lines.join("\n");
}

function queryValidate(config: TerraformConfig): string {
  const issues: string[] = [];

  for (const block of config.blocks.values()) {
    if (block.kind === "resource" && block.attributes.size === 0 && block.nestedBlocks.length === 0) {
      issues.push(`  WARNING: resource ${block.fullType}.${block.label} has no attributes`);
    }
    if (block.kind === "output" && !block.attributes.has("value")) {
      issues.push(`  ERROR: output "${block.label}" missing required "value" attribute`);
    }
    if (block.kind === "module" && !block.attributes.has("source")) {
      issues.push(`  ERROR: module "${block.label}" missing required "source" attribute`);
    }
  }

  // Check for dangling connections
  for (const conn of config.connections.values()) {
    if (!config.blocks.has(conn.sourceId)) {
      issues.push(`  ERROR: connection references missing source block`);
    }
    if (!config.blocks.has(conn.targetId)) {
      issues.push(`  ERROR: connection references missing target block`);
    }
  }

  if (issues.length === 0) return "Configuration is valid.";
  return `Found ${issues.length} issue(s):\n${issues.join("\n")}`;
}

function queryStats(config: TerraformConfig): string {
  const counts = new Map<string, number>();
  for (const block of config.blocks.values()) {
    counts.set(block.kind, (counts.get(block.kind) ?? 0) + 1);
  }

  const lines: string[] = [`Total blocks: ${config.blocks.size}`];
  for (const [kind, count] of counts) {
    lines.push(`  ${kind}: ${count}`);
  }
  lines.push(`Connections: ${config.connections.size}`);
  return lines.join("\n");
}

function queryStatus(config: TerraformConfig): string {
  const lines: string[] = [];
  lines.push(`Title: ${config.title}`);
  lines.push(`File: ${config.filePath ?? "(unsaved)"}`);
  lines.push(`Blocks: ${config.blocks.size}`);
  lines.push(`Connections: ${config.connections.size}`);
  return lines.join("\n");
}

function queryFind(config: TerraformConfig, text: string): string {
  if (!text) return "find requires search text";
  const lower = text.toLowerCase();
  const matches = [...config.blocks.values()].filter(
    (b) => b.label.toLowerCase().includes(lower) || b.fullType.toLowerCase().includes(lower),
  );
  if (matches.length === 0) return `No blocks matching "${text}"`;
  return matches.map((b) => `  ${b.kind} ${b.fullType}.${b.label}`).join("\n");
}

function queryHistory(eventLog: EventLog<TerraformEvent> | undefined, count: number): string {
  if (!eventLog) return "No event log available.";
  const events = eventLog.recent(count);
  if (events.length === 0) return "No events.";

  return events.map((e, i) => {
    switch (e.type) {
      case "block_added": return `  ${i + 1}. + ${e.block.kind} ${e.block.label}`;
      case "block_removed": return `  ${i + 1}. - ${e.block.kind} ${e.block.label}`;
      case "attribute_set": return `  ${i + 1}. * set ${e.key} on block`;
      case "attribute_removed": return `  ${i + 1}. * unset ${e.key}`;
      case "connection_added": return `  ${i + 1}. ~ ${e.connection.sourceLabel} -> ${e.connection.targetLabel}`;
      case "connection_removed": return `  ${i + 1}. - disconnect`;
      case "tag_set": return `  ${i + 1}. * tag ${e.key}`;
      case "tag_removed": return `  ${i + 1}. - tag ${e.key}`;
      case "block_renamed": return `  ${i + 1}. * rename ${e.before} → ${e.after}`;
      default: return `  ${i + 1}. ${e.type}`;
    }
  }).join("\n");
}

File: internal/terraform/queries.go
package terraform

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/os-tack/fcp-terraform/internal/fcpcore"
)

// DispatchQuery routes a query string to the appropriate handler.
func DispatchQuery(query string, model *TerraformModel, eventLog *fcpcore.EventLog) string {
	parts := strings.Fields(query)
	if len(parts) == 0 {
		return unknownQuery("")
	}

	cmd := strings.ToLower(parts[0])
	args := parts[1:]

	switch cmd {
	case "plan":
		return queryPlan(model)
	case "graph":
		return queryGraph(model)
	case "describe":
		return queryDescribe(model, args)
	case "stats":
		return queryStats(model)
	case "map":
		return queryMap(model)
	case "status":
		return queryStatus(model)
	case "history":
		return queryHistory(eventLog, args)
	case "list":
		return queryList(model, args)
	case "find":
		return queryFind(model, strings.Join(args, " "))
	default:
		return unknownQuery(cmd)
	}
}

func unknownQuery(cmd string) string {
	available := "plan, graph, describe, stats, map, status, history, list, find"
	if cmd == "" {
		return "Query required. Available: " + available
	}
	return fmt.Sprintf("Unknown query %q. Available: %s", cmd, available)
}

// ── plan ─────────────────────────────────────────────────

func queryPlan(model *TerraformModel) string {
	b := model.Bytes()
	if len(strings.TrimSpace(string(b))) == 0 {
		return "Empty configuration. Add some resources first."
	}
	return string(b)
}

// ── graph ────────────────────────────────────────────────

func queryGraph(model *TerraformModel) string {
	if len(model.Index.Connections) == 0 {
		return "No connections. Use 'connect SRC -> TGT' to add dependencies."
	}

	lines := []string{"Dependency Graph:"}
	for src, targets := range model.Index.Connections {
		for dst, edgeLabel := range targets {
			suffix := ""
			if edgeLabel != "" {
				suffix = " (" + edgeLabel + ")"
			}
			lines = append(lines, fmt.Sprintf("  %s -> %s%s", src, dst, suffix))
		}
	}
	return strings.Join(lines, "\n")
}

// ── describe ─────────────────────────────────────────────

func queryDescribe(model *TerraformModel, args []string) string {
	if len(args) == 0 {
		return "describe requires a LABEL"
	}

	label := args[0]
	ref := model.resolveRef(label)
	if ref == nil {
		return fmt.Sprintf("block %q not found", label)
	}

	lines := []string{}

	// Header
	var header string
	if ref.Kind == "resource" || ref.Kind == "data" {
		header = fmt.Sprintf("%s %s.%s", ref.Kind, ref.FullType, ref.Label)
	} else {
		header = fmt.Sprintf("%s %q", ref.Kind, ref.Label)
	}
	lines = append(lines, header)

	if ref.Provider != "" && ref.Kind != "provider" {
		lines = append(lines, fmt.Sprintf("  provider: %s", ref.Provider))
	}

	// Attributes from hclwrite
	body := ref.Block.Body()
	attrs := body.Attributes()
	if len(attrs) > 0 {
		lines = append(lines, "  attributes:")
		for name, attr := range attrs {
			valBytes := strings.TrimSpace(string(attr.Expr().BuildTokens(nil).Bytes()))
			lines = append(lines, fmt.Sprintf("    %s = %s", name, valBytes))
		}
	}

	// Tags
	if len(ref.Tags) > 0 {
		lines = append(lines, "  tags:")
		for k, v := range ref.Tags {
			lines = append(lines, fmt.Sprintf("    %s = %q", k, v))
		}
	}

	// Nested blocks
	nested := body.Blocks()
	if len(nested) > 0 {
		lines = append(lines, "  nested blocks:")
		for _, nb := range nested {
			lines = append(lines, fmt.Sprintf("    %s {}", nb.Type()))
			for name, attr := range nb.Body().Attributes() {
				valBytes := strings.TrimSpace(string(attr.Expr().BuildTokens(nil).Bytes()))
				lines = append(lines, fmt.Sprintf("      %s = %s", name, valBytes))
			}
		}
	}

	// Connections
	srcKey := strings.ToLower(ref.Label)
	outgoing := model.Index.Connections[srcKey]
	var incoming []string
	for src, targets := range model.Index.Connections {
		if _, ok := targets[srcKey]; ok && src != srcKey {
			incoming = append(incoming, src)
		}
	}

	if len(outgoing) > 0 || len(incoming) > 0 {
		lines = append(lines, "  connections:")
		for dst := range outgoing {
			lines = append(lines, fmt.Sprintf("    -> %s", dst))
		}
		for _, src := range incoming {
			lines = append(lines, fmt.Sprintf("    <- %s", src))
		}
	}

	return strings.Join(lines, "\n")
}

// ── stats ────────────────────────────────────────────────

func queryStats(model *TerraformModel) string {
	counts := make(map[string]int)
	for _, label := range model.Index.Order {
		ref := model.Index.Get(label)
		if ref != nil {
			counts[ref.Kind]++
		}
	}

	total := len(model.Index.Order)
	lines := []string{fmt.Sprintf("Total blocks: %d", total)}

	// Show counts by kind in a consistent order
	for _, kind := range []string{"resource", "data", "variable", "output", "provider", "module"} {
		if c, ok := counts[kind]; ok {
			lines = append(lines, fmt.Sprintf("  %s: %d", kind, c))
		}
	}

	// Unique providers
	if len(model.Index.ByProvider) > 0 {
		providers := make([]string, 0, len(model.Index.ByProvider))
		for p := range model.Index.ByProvider {
			providers = append(providers, p)
		}
		lines = append(lines, fmt.Sprintf("Providers: %s", strings.Join(providers, ", ")))
	}

	connCount := 0
	for _, targets := range model.Index.Connections {
		connCount += len(targets)
	}
	lines = append(lines, fmt.Sprintf("Connections: %d", connCount))

	return strings.Join(lines, "\n")
}

// ── map ──────────────────────────────────────────────────

func queryMap(model *TerraformModel) string {
	lines := []string{fmt.Sprintf("Terraform Config: %s", model.Title), ""}

	providers := model.Index.FindByKind("provider")
	resources := model.Index.FindByKind("resource")
	dataSources := model.Index.FindByKind("data")
	variables := model.Index.FindByKind("variable")
	outputs := model.Index.FindByKind("output")
	modules := model.Index.FindByKind("module")

	if len(providers) > 0 {
		names := make([]string, len(providers))
		for i, p := range providers {
			names[i] = p.Label
		}
		lines = append(lines, fmt.Sprintf("Providers: %s", strings.Join(names, ", ")))
	}

	if len(resources) > 0 {
		typeCounts := make(map[string]int)
		for _, r := range resources {
			typeCounts[r.FullType]++
		}
		parts := make([]string, 0, len(typeCounts))
		for t, c := range typeCounts {
			parts = append(parts, fmt.Sprintf("%s x%d", t, c))
		}
		lines = append(lines, fmt.Sprintf("Resources (%d): %s", len(resources), strings.Join(parts, ", ")))
	}

	if len(dataSources) > 0 {
		lines = append(lines, fmt.Sprintf("Data Sources: %d", len(dataSources)))
	}
	if len(variables) > 0 {
		lines = append(lines, fmt.Sprintf("Variables: %d", len(variables)))
	}
	if len(outputs) > 0 {
		lines = append(lines, fmt.Sprintf("Outputs: %d", len(outputs)))
	}
	if len(modules) > 0 {
		lines = append(lines, fmt.Sprintf("Modules: %d", len(modules)))
	}

	connCount := 0
	for _, targets := range model.Index.Connections {
		connCount += len(targets)
	}
	if connCount > 0 {
		lines = append(lines, fmt.Sprintf("Connections: %d", connCount))
	}

	return strings.Join(lines, "\n")
}

// ── status ───────────────────────────────────────────────

func queryStatus(model *TerraformModel) string {
	filePath := model.FilePath
	if filePath == "" {
		filePath = "(unsaved)"
	}

	lines := []string{
		fmt.Sprintf("Title: %s", model.Title),
		fmt.Sprintf("File: %s", filePath),
		fmt.Sprintf("Blocks: %d", len(model.Index.Order)),
	}

	connCount := 0
	for _, targets := range model.Index.Connections {
		connCount += len(targets)
	}
	lines = append(lines, fmt.Sprintf("Connections: %d", connCount))

	return strings.Join(lines, "\n")
}

// ── history ──────────────────────────────────────────────

func queryHistory(eventLog *fcpcore.EventLog, args []string) string {
	if eventLog == nil {
		return "No event log available."
	}

	count := 10
	if len(args) > 0 {
		if n, err := strconv.Atoi(args[0]); err == nil && n > 0 {
			count = n
		}
	}

	events := eventLog.Recent(count)
	if len(events) == 0 {
		return "No events."
	}

	lines := make([]string, len(events))
	for i, ev := range events {
		// Events are SnapshotEvents with Summary strings
		if se, ok := ev.(*SnapshotEvent); ok {
			lines[i] = fmt.Sprintf("  %d. %s", i+1, se.Summary)
		} else if s, ok := ev.(string); ok {
			lines[i] = fmt.Sprintf("  %d. %s", i+1, s)
		} else {
			lines[i] = fmt.Sprintf("  %d. %v", i+1, ev)
		}
	}
	return strings.Join(lines, "\n")
}

// ── list ─────────────────────────────────────────────────

func queryList(model *TerraformModel, args []string) string {
	var refs []*BlockRef

	// Filter by selector if provided
	if len(args) > 0 && strings.HasPrefix(args[0], "@") {
		sel := ParseSelector(args[0])
		refs = ResolveSelector(sel, model)
	} else {
		// All blocks in insertion order
		for _, label := range model.Index.Order {
			if ref := model.Index.Get(label); ref != nil {
				refs = append(refs, ref)
			}
		}
	}

	if len(refs) == 0 {
		return "No blocks found."
	}

	lines := make([]string, len(refs))
	for i, ref := range refs {
		qualified := ref.QualifiedName()
		attrCount := len(ref.Block.Body().Attributes())
		lines[i] = fmt.Sprintf("  %-10s %-35s (%d attrs)", ref.Kind, qualified, attrCount)
	}
	return strings.Join(lines, "\n")
}

// ── find ─────────────────────────────────────────────────

func queryFind(model *TerraformModel, text string) string {
	if text == "" {
		return "find requires search text"
	}

	lower := strings.ToLower(text)
	var matches []*BlockRef
	for _, label := range model.Index.Order {
		ref := model.Index.Get(label)
		if ref == nil {
			continue
		}
		if strings.Contains(strings.ToLower(ref.Label), lower) ||
			strings.Contains(strings.ToLower(ref.FullType), lower) {
			matches = append(matches, ref)
		}
	}

	if len(matches) == 0 {
		return fmt.Sprintf("No blocks matching %q", text)
	}

	lines := make([]string, len(matches))
	for i, ref := range matches {
		lines[i] = fmt.Sprintf("  %s %s", ref.Kind, ref.QualifiedName())
	}
	return strings.Join(lines, "\n")
}
